src/vs/workbench/services/search/node/textSearch.ts

```diff
@@ -5,330 +5,178 @@
 	}
 
-	public cancel(): void {
+	cancel(): void {
 		this.isCanceled = true;
 		this.walker.cancel();
-	}
 
-	public search(onResult: (match: ISerializedFileMatch) => void, onProgress: (progress: IProgress) => void, done: (error: Error, complete: ISerializedSearchComplete) => void): void {
-		let resultCounter = 0;
+		this.workers.forEach(w => w.cancel());
+	}
 
-		let progress = () => {
-			this.progressed++;
-			if (this.progressed % Engine.PROGRESS_FLUSH_CHUNK_SIZE === 0) {
-				onProgress({ total: this.total, worked: this.worked }); // buffer progress in chunks to reduce pressure
+	search(onResult: (match: ISerializedFileMatch) => void, onProgress: (progress: IProgress) => void, done: (error: Error, complete: ISerializedSearchComplete) => void): void {
+		this.startWorkers();
+		const progress = () => {
+			if (++this.progressed % Engine.PROGRESS_FLUSH_CHUNK_SIZE === 0) {
+				onProgress({ total: this.totalBytes, worked: this.processedBytes }); // buffer progress in chunks to reduce pressure
 			}
 		};
 
-		let unwind = (processed: number) => {
-			this.worked += processed;
+		const unwind = (processed: number) => {
+			this.processedBytes += processed;
 
 			// Emit progress() unless we got canceled or hit the limit
 			if (processed && !this.isDone && !this.isCanceled && !this.limitReached) {
 				progress();
 			}
 
 			// Emit done()
-			if (this.worked === this.total && this.walkerIsDone && !this.isDone) {
+			if (!this.isDone && this.processedBytes === this.totalBytes && this.walkerIsDone) {
 				this.isDone = true;
+				this.disposeWorkers();
 				done(this.walkerError, {
 					limitHit: this.limitReached,
 					stats: this.walker.getStats()
 				});
 			}
 		};
 
-		// Walk over the file system
-		this.walker.walk(this.rootFolders, this.extraFiles, result => {
-			const size = result.size || 1;
-			this.total += size;
-
-			// If the result is empty or we have reached the limit or we are canceled, ignore it
-			if (this.limitReached || this.isCanceled) {
-				return unwind(size);
-			}
+		const run = (batch: string[], batchBytes: number): void => {
+			const worker = this.workers[this.nextWorker];
+			this.nextWorker = (this.nextWorker + 1) % this.workers.length;
 
-			// Indicate progress to the outside
-			progress();
-
-			let fileMatch: FileMatch = null;
-
-			let doneCallback = (error?: Error) => {
-				if (!error && !this.isCanceled && fileMatch && !fileMatch.isEmpty()) {
-					onResult(fileMatch.serialize());
-				}
-
-				return unwind(size);
-			};
-
-			const absolutePath = result.base ? [result.base, result.relativePath].join(path.sep) : result.relativePath;
-			let perLineCallback = (line: string, lineNumber: number) => {
-				if (this.limitReached || this.isCanceled) {
-					return; // return early if canceled or limit reached
+			const maxResults = this.config.maxResults - this.numResults;
+			worker.search({ absolutePaths: batch, maxResults }).then(result => {
```

Comment:
Previously we didn't surface errors from the text search, just from the file walker. I wasn't sure whether to change that. I guess I should do something with the error.